### PR TITLE
use VK_EXT_physical_device_drm to ensure drm and vk open the same GPU

### LINKED ISSUE #18: DRM and Vulkan should open the same GPU
Right now for simplicity DRM and Vulkan open the GPU independently. This is a problem if the system has more than one GPU, because they might open different GPUs. I don't think it will ever make sense to have compositing GPU != displaying GPU. There are two possibilities:

1.  Open with DRM and tell Vulkan our choice
I think this is doable with mesa's recently added `device-select-layer`

2. Open with Vulkan and tell DRM our choice
This is not possible in core Vulkan, there is no way to extract the file descriptor or device node. However, the extension `VK_EXT_pci_bus_info` might provide enough info to be able to so.
EDIT: there will be a new extension `VK_EXT_physical_device_drm`

### PR DESCRIPTION
**Pull request by Ella-0**
closes #18

### CODE REVIEW

**Comment by Ella-0** [comment]
Weirdly I can't get this to work. I'm getting `/dev/dri/card0` fine just always failing with `The display is already taken by another program`. Can anyone else test?

**Comment by st3r4g** [comment]
Enabling `VK_KHR_display` makes vulkan open the primary node, which maybe prevents us to open it again. I wonder if it's ok to assume that the corresponding primary minor is render minor - 128. That wouldn't require enabling `VK_KHR_display`.

**Comment by Ella-0** [comment]
Might be different on the *bsds but I'm not sure. Perhaps patching mesa is a better option though I'm not sure. I'll see what the `VK_EXT_physical_device_drm` spec says

**Comment by Ella-0** [comment]
Just taken a look at the implementations upstream and radv seams like it should set the primary device stuff up even if `VK_KHR_display` isn't enabled. Unfortunately I don't have the required hardware to test that.

**Comment by Ella-0** [comment]
One way around this would be to open the device with drm first, then filter through available Vulkan devices and compare their dev numbers until one matches.

**Comment by st3r4g** [comment]
> prevents us to open it again

to be precise, we can open it, but not become master, which is probably tied to the file description.

> Perhaps patching mesa is a better option

yeah, imo it should work also without `VK_KHR_display`. Let me check if it works on radv...

**Comment by st3r4g** [comment]
> Let me check if it works on radv...

It does. Let's not add `VK_KHR_display` and use the "render minor - 128" temporary fallback (when `hasPrimary == 0`) until anv has a proper implementation.

**Comment by Ella-0** [comment]
Well I can make an MR to fix the anv implementation right now.

**Comment by Ella-0** [comment]
[https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/11616](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/11616)

**Comment by Ella-0** [comment]
v3dv might be a little more complicated because an authenticated primary node is required from the compositor in order for the driver to function. (though this is likely a bug somewhere, upstream still hasn't merged `VK_EXT_physical_device_drm` and investigating this weird issue; I encountered it when implementing Wayland WSI support in v3dv https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/broadcom/vulkan/v3dv_device.c#L431)

**Comment by st3r4g** [comment]
> https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/11616

yeah, that works. But the `has_master` variable name is a bit confusing imo, seems to imply you have master, while it's just the existence of the primary node.

**Comment by Ella-0** [comment]
Yeah it is a bit, I named it that to match the naming of file descriptors in the driver.

**Comment by Ella-0** [comment]
That's been merged now.

File: src/backend/vulkan.c [hunk 1]
```diff
@@ -152,18 +153,28 @@ VkPhysicalDevice get_physical_device(VkInstance inst) {
...
 	boxlog("Device name: %s", props.deviceName);
 	boxlog("       type: %s", type_to_string(props.deviceType));
-/*	errlog("%d %d %d %d", busProps.pciDomain, busProps.pciBus,
-	 busProps.pciDevice, busProps.pciFunction);*/
+	if (drm_props.hasPrimary) {
+		boxlog("    primary: %d:%d", drm_props.primaryMajor, drm_props.primaryMinor);
+		*major = drm_props.primaryMajor;
+		*minor = drm_props.primaryMinor;
+	} else {
+    		errlog("failed to get primary device");
```
Comment (st3r4g) [inline src/backend/vulkan.c]: If there is no primary node associated to the device, we should abort (we can't do modesetting). Let's add `return VK_NULL_HANDLE;` here.
Later we need to add some proper logic for the device selection, right now it is broken if there is more than 1 gpu.

Comment (Ella-0) [reply src/backend/vulkan.c]: Done that. Unfortunately my only dual GPU system has an nvidia card :confused:

**Comment by st3r4g** [comment]
Wait a second, why is it a *device* extension? We need to use it before creating a device... (a device is created from the physical device). How do we check/enable it before using it?? Should've been an *instance* extension, no?

**Comment by Ella-0** [comment]
Yeah I was wondering that. I thought it was an instance extension but it isn't. I suppose it's because you don't necessarily know what physical device ur using at instance creation time and not all physical devices will support it.

**Comment by st3r4g** [comment]
> We need to use it before creating a device

Not really, we could as well move it at the end of `create_device`. We lose a potential check we could have made at device selection, but not a big deal. I can do it tomorrow before merging.

**Comment by st3r4g** [comment]
Thanks for this one and for fixing anv!

**Comment by Ella-0** [comment]
np (◕ᴗ◕✿)

**Pull request merged**